modules/chapter1/pages/section2.adoc

```diff
@@ -1,3 +1,355 @@
+      "name": "random-num-2-Output", "path": "/tmp/outputs/Output/data"}]}'
+    tekton.dev/input_artifacts: '{"print-msg": [{"name": "random-num-Output", "parent_task":
+      "random-num"}], "print-msg-2": [{"name": "random-num-Output", "parent_task":
+      "random-num"}], "print-msg-3": [{"name": "random-num-2-Output", "parent_task":
+      "random-num-2"}], "print-msg-4": [{"name": "random-num-2-Output", "parent_task":
+      "random-num-2"}]}'
+    tekton.dev/artifact_bucket: mlpipeline
+    tekton.dev/artifact_endpoint: minio-service.kubeflow:9000
+    tekton.dev/artifact_endpoint_scheme: http://
+    tekton.dev/artifact_items: '{"flip-coin": [["Output", "$(results.Output.path)"]],
+      "print-msg": [], "print-msg-2": [], "print-msg-3": [], "print-msg-4": [], "random-num":
+      [["Output", "$(results.Output.path)"]], "random-num-2": [["Output", "$(results.Output.path)"]]}'
+    sidecar.istio.io/inject: "false"
+    tekton.dev/template: ''
+----
+****
+
+The resulting yaml file _(coin-toss.yaml)_ can then be uploaded throught the UI
+****
+image::import-pipeline.png[]
+****
+****
+image::pipeline-imported.png[]
+****
+Once imported the structure of the _DAG_ will be shown. Each step in the pipeline will be run as a container on OpenShift.
+
+****
+image::pipeline-run-view.png[]
+****
+
+To execute the pipeline, click on _Create Run_ in the menu and fill out the _Name_ and _Description_.
+If the pipeline has _Input Parameters_ or a you need to schedule a recurring run then that can be configured further down. Once ready clink _Create_ and the pipeline will be scheduled.
+
+****
+image::creating-pipeline-run.png[]
+****
+
+The pipeline will execute and the outputs will be stored into the configured S3 bucket.
+As the pipeline executes the view will be updated to show the steps being executed. It's possible to click on the graph nodes to reveal information of the steps
+
+****
+image::post-pipeline-run.png[]
+****
+
+Once the pipeline has completed it is possible to access the output and pipeline artifacts (if used) in the Minio Storage UI
+
+****
+image::object-store-after-run.png[]
+****
+
+== Advanced Example
+
+In this section we're going to demonstrate how Data Science Pipelines work in a real world scenario. 
+
+****
+In this scenario we have a remote edge device which uses an AI model to manage the characteristics of its battery usage depending on the environment it's deployed in. On a regular schedule it uploads battery events via a data gateway and those batter events are used to train a model which is then retrieved by the device and used. 
+
+
+image::openshift-ai-dsp-edge.png[]
+****
+
+The entire pipeline is here link:sample-pipeline-full.py[Pipeline]
+
+We're not going to go through all of it but focus on the key aspects of it. 
+
+The actual pipeline is defined by the following function
+
+****
+[source,python]
+include::sample-pipeline-full.py[lines=450..454]
```

Comment:
This python snippet is not being rendered 

![image](https://github.com/RedHatQuickCourses/rhods-pipelines/assets/675470/44b8b292-23a7-4def-aea5-d9bab2713659)

Reply:
Same problem occurs in the rest of python snippets below.

Reply:
yep I've fixed the rendering issue. I don't think a workbench needs to be created to use a DSP